modules/core/src/main/java/org/apache/synapse/util/InlineExpressionUtil.java

```diff
@@ -203,42 +204,53 @@ private static boolean isValidXML(String stringToValidate) {
      * Inline expressions will be denoted inside ${}
      * e.g.: ${var.var1}, ${payload.element.id}
      *
-     * @param inlineText Inline text string
+     * @param inlineText      Inline text string
+     * @param expressionCache Cache to store the synapse expressions
      * @return true if the string contains content aware inline synapse expressions, false otherwise
      */
-    public static boolean isInlineSynapseExpressionsContentAware(String inlineText) {
+    public static boolean initInlineSynapseExpressions(String inlineText, Map<String, SynapseExpression> expressionCache)
+            throws JaxenException {
 
+        boolean isContentAware = false;
         Matcher matcher = SYNAPSE_EXPRESSION_PLACEHOLDER_PATTERN.matcher(inlineText);
         while (matcher.find()) {
-            // Extract the expression inside ${...}
-            String expression = matcher.group(1);
-            if (SynapseExpressionUtils.isSynapseExpressionContentAware(expression)) {
-                return true;
+            // Extract the expression inside ${...} and add it to the cache
+            String placeholder = matcher.group(1);
+            SynapseExpression expression = new SynapseExpression(placeholder);
+            if (expression.isContentAware()) {
+                isContentAware = true;
             }
+            expressionCache.put(placeholder, expression);
         }
-        return false;
+        return isContentAware;
     }
 
     /**
      * Process the inline template and replace the synapse expressions with the resolved values
      *
-     * @param synCtx   Message Context
-     * @param template Inline template
+     * @param synCtx          Message Context
+     * @param template        Inline template
+     * @param expressionCache Cache to store the synapse expressions
      * @return Processed inline template
      */
-    public static String processInLineSynapseExpressionTemplate(MessageContext synCtx, String template)
+    public static String processInLineSynapseExpressionTemplate(MessageContext synCtx, String template,
+                                                                Map<String, SynapseExpression> expressionCache)
             throws JaxenException {
 
         Matcher matcher = SYNAPSE_EXPRESSION_PLACEHOLDER_PATTERN.matcher(template);
         StringBuffer result = new StringBuffer();
         while (matcher.find()) {
             // Extract the expression inside ${...}
             String placeholder = matcher.group(1);
-            SynapseExpression expression = new SynapseExpression(placeholder);
+            SynapseExpression expression = expressionCache.get(placeholder);
+            if (expression == null) {
+                expression = new SynapseExpression(placeholder);
+                expressionCache.put(placeholder, expression);
+            }
             String replacement = expression.stringValueOf(synCtx);
```

Comment:
shall we calculate replacement values for the entire cached map at once. If we have the same expression used in several places, we don't have to evaluate again.

Reply:
Fixed with c699634